### PR TITLE
VideoCapture segfault fix for #22126

### LINKED ISSUE #22126: cv2.VideoCapture segfaults
OpenCV 4.6, Centos7

This file:  [cv2-crash.zip](https://github.com/opencv/opencv/files/8933393/cv2-crash.zip)

Testcase:
````
# pip3 install opencv-contrib-python-headless
import cv2
cap = cv2.VideoCapture('cv2-crash.mkv')
while True:
    ret, frame = cap.read()
    if not ret:
        break
    cap_time = cap.get(cv2.CAP_PROP_POS_MSEC)
    print( cap_time )
cap.release()
````

Console:
````
[...]
163992.0
Core Dump (Segmentation fault)
````

### PR DESCRIPTION
**Pull request by danopdev**
resolves #22126

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ x ] I agree to contribute to the project under Apache 2 License.
- [ x ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ x ] The PR is proposed to the proper branch
- [ x ] There is a reference to the original bug report and related work
- [ x ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ x ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by alalek — CHANGES_REQUESTED** [review]
Please run reproducer locally first. Fix is not valid.

**Comment by danopdev** [comment]
I used this code for testing:

```
#include <opencv2/videoio.hpp>

using namespace cv;

int main() {
    VideoCapture vc = VideoCapture("cv2-crash.mkv");
    Mat frame;

    while(vc.read(frame)) {
        double value = vc.get(cv::CAP_PROP_POS_MSEC);
        printf("pos: %f\n", value);
    }

    vc.release();

    return 0;
}
```

The video is the one attached to the issue (https://github.com/opencv/opencv/issues/22126).

Before the patch (original 4.x) when I run the application it ends with:
```
pos: 163659.000000
pos: 163692.000000
pos: 163725.000000
pos: 163759.000000
pos: 163792.000000
pos: 163959.000000
pos: 163992.000000
Segmentation fault
```

After the patch:
```
pos: 179258.000000
pos: 179358.000000
pos: 179458.000000
pos: 179558.000000
pos: 179658.000000
pos: 179758.000000
pos: 0.000000
pos: 0.000000
```
(no segfault).

@[alalek](https://github.com/alalek): Can you please provide me a link to a video that don't work so I can test it ? 

NOTE: I rebased the code to have only one commit.

**Comment by alalek** [comment]
Check content of returned frame data (use video from the issue). Frame data is corrupted.

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 1]
```diff
@@ -1633,7 +1633,7 @@ bool CvCapture_FFMPEG::retrieveFrame(int flag, unsigned char** data, int* step,
             img_convert_ctx,
             sw_picture->data,
             sw_picture->linesize,
-            0, context->coded_height,
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp:1636]: Lets use `sw_picture->height` here.

According to ffmpeg sample code: https://github.com/FFmpeg/FFmpeg/blob/n4.4.2/fftools/ffplay.c#L925-L926

---

BTW, from the same samples code there is no usage of `coded_width`/`coded_height` values with modern FFmpeg.
Perhaps we need to update `sws_getCachedContext()` call too (probably later, and align with "n0,8" ffmpeg version).

**Review by alalek — COMMENTED** [review]
Looks like it works. Thank you!

As a bug fix this patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into 4.x regularly (weekly/bi-weekly).

Please:
- change "base" branch of this PR: 4.x => 3.4 (use "Edit" button near PR title)
- rebase your commits from 4.x onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/4.x`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

File: modules/videoio/src/cap_ffmpeg_impl.hpp (hunk 1 above)
Comment (danopdev) [reply modules/videoio/src/cap_ffmpeg_impl.hpp:1636]: Actually I tried 3.4 as it is and it works like a charm. No segfault.
Even the issue https://github.com/opencv/opencv/pull/22145 is not present.
Maybe there are some specific 4.x changes that brakes it.

Comment (alalek) [reply modules/videoio/src/cap_ffmpeg_impl.hpp:1636]: Code is different, but there is the same problem on 3.4: https://github.com/opencv/opencv/blob/3.4.18/modules/videoio/src/cap_ffmpeg_impl.hpp#L1409-L1411

Comment (danopdev) [reply modules/videoio/src/cap_ffmpeg_impl.hpp:1636]: My bad. I don't know why but from time to time it pass without crashing.
But 3.4 can crash too.
I'll update the PR.

Comment (danopdev) [reply modules/videoio/src/cap_ffmpeg_impl.hpp]: PR ready for 3.4 (builds and run correctly)

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 2]
```diff
@@ -1408,7 +1408,7 @@ bool CvCapture_FFMPEG::retrieveFrame(int, unsigned char** data, int* step, int*
             img_convert_ctx,
             picture->data,
             picture->linesize,
-            0, video_st->codec->coded_height,
+            0, video_st->codec->height,
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp:1411]: Again, please use `picture->height` here.

According to ffmpeg sample code: https://github.com/FFmpeg/FFmpeg/blob/n4.4.2/fftools/ffplay.c#L925-L926

Comment (danopdev) [reply modules/videoio/src/cap_ffmpeg_impl.hpp]: You are right, we use picture->data and picture->linesize so it make sense.
Done (build and run OK)

**Review by alalek — APPROVED** [review]
Thank you!

**Comment by chercy1** [comment]
> resolves #22126
> 
> ### Pull Request Readiness Checklist
> See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request
> 
> * [ x ] I agree to contribute to the project under Apache 2 License.
> * [ x ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
> * [ x ] The PR is proposed to the proper branch
> * [ x ] There is a reference to the original bug report and related work
> * [ x ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
>   Patch to opencv_extra has the same branch name.
> * [ x ] The feature is well documented and sample code can be built with the project CMake

Can you explain the reason for this problem? What caused cv2.VideoCapture segfaults?

**Pull request merged**